File: tracks/java/src/triangle/Triangle.java
package triangle;

/**
 * The Triangle class represents a triangle defined by its three sides.
 * It provides methods to determine if the triangle is equilateral, isosceles,
 * or scalene.
 * The class also includes a constructor that validates the sides of the
 * triangle,
 * ensuring they adhere to the triangle inequality theorem.
 */
public class Triangle {
    /**
     * Length of the first side of the triangle.
     */
    private double side1;

    /**
     * Length of the second side of the triangle.
     */
    private double side2;

    /**
     * Length of the third side of the triangle.
     */
    private double side3;

    /**
     * Constructs a Triangle object with the specified side lengths.
     * Validates the sides to ensure they can form a valid triangle according to
     * the triangle inequality theorem.
     * 
     * @param side1 Length of the first side of the triangle.
     * @param side2 Length of the second side of the triangle.
     * @param side3 Length of the third side of the triangle.
     * @throws TriangleException if the sides do not form a valid triangle.
     */
    Triangle(double side1, double side2, double side3) throws TriangleException {
        if (side1 <= 0 || side2 <= 0 || side3 <= 0 || side1 + side2 < side3 || side2 + side3 < side1
                || side1 + side3 < side2) {
            throw new TriangleException();
        }

        this.side1 = side1;
        this.side2 = side2;
        this.side3 = side3;
    }

    /**
     * Determines if the triangle is equilateral (all sides are equal).
     * 
     * @return true if the triangle is equilateral, false otherwise.
     */
    boolean isEquilateral() {
        return this.side1 == this.side2 && this.side2 == this.side3;
    }

    /**
     * Determines if the triangle is isosceles (two sides are equal).
     * 
     * @return true if the triangle is isosceles, false otherwise.
     */
    boolean isIsosceles() {
        return (this.side1 == this.side2) || (this.side2 == this.side3) || (this.side3 == this.side1);
    }

    /**
     * Determines if the triangle is scalene (no sides are equal).
     * 
     * @return true if the triangle is scalene, false otherwise.
     */
    boolean isScalene() {
        return this.side1 != this.side2 && this.side2 != this.side3 && this.side3 != this.side1;
    }
}

File: tracks/python/triangle.py
"""
Triangle Type Determination Module.

This module provides functions to determine the type of a triangle based on the lengths of its sides.
It includes checks for the validity of a triangle and determines whether it is equilateral, isosceles, or scalene.

Functions:
- triangle(sides): Determines if the given sides can form a valid triangle.
- equilateral(sides): Checks if a triangle is equilateral (all sides equal).
- isosceles(sides): Checks if a triangle is isosceles (at least two sides equal).
- scalene(sides): Checks if a triangle is scalene (all sides different).
"""


def triangle(sides):
    """
    Determines if the given sides can form a valid triangle.

    :param sides: list of int or float - lengths of the three sides of a potential triangle.
    :return: bool - True if the sides can form a triangle, False otherwise.
    """
    if sides[0] > 0 and sides[1] > 0 and sides[2] > 0:
        if (sides[0] + sides[1]) >= sides[2] and (sides[1] + sides[2]) >= sides[0] and (sides[0] + sides[2]) >= sides[1]:
            return True
    return False


def equilateral(sides):
    """
    Checks if a triangle is equilateral, meaning all sides are equal.

    :param sides: list of int or float - lengths of the three sides of the triangle.
    :return: bool - True if the triangle is equilateral, False otherwise.
    """
    return triangle(sides) and sides[0] == sides[1] and sides[1] == sides[2]


def isosceles(sides):
    """
    Checks if a triangle is isosceles, meaning at least two sides are equal.

    :param sides: list of int or float - lengths of the three sides of the triangle.
    :return: bool - True if the triangle is isosceles, False otherwise.
    """
    return triangle(sides) and (sides[0] == sides[1] or sides[1] == sides[2] or sides[0] == sides[2])


def scalene(sides):
    """
    Checks if a triangle is scalene, meaning all sides are different.

    :param sides: list of int or float - lengths of the three sides of the triangle.
    :return: bool - True if the triangle is scalene, False otherwise.
    """
    return triangle(sides) and (sides[0] != sides[1] and sides[1] != sides[2] and sides[0] != sides[2])
